Implement builtin tool filtering

This should work

```yaml
version: "2"

agents:
  root:
    instruction: help the user
    model: openai/gpt-5
    toolsets:
      - type: filesystem
        tools: [list_allowed_directories]
```

Today we have a panic since the tools are half-filtered.

This manifest itself by 
```
list_allowed_directories()
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x116061b]

goroutine 83 [running]:
github.com/docker/cagent/pkg/runtime.(*runtime).runTool(0xc00067e480, {0x1ccb7e8, 0xc000282d20}, {0xc000168b40?, 0x0?}, {{0xc00087a789, 0x1e}, {0x1829c8b, 0x8}, {{0xc00087a7b1, ...}, ...}}, ...)
```

--- a/pkg/tools/builtin/filesystem.go
+++ b/pkg/tools/builtin/filesystem.go
@@ -10,6 +10,7 @@ import (
 	"os/exec"
 	"path/filepath"
 	"regexp"
+	"slices"
 	"strings"
 	"time"
 
@@ -443,14 +444,12 @@ func (t *FilesystemTool) Tools(context.Context) ([]tools.Tool, error) {
 	}
 
 	var allowedTools []tools.Tool
-	for _, tool := range t.allowedTools {
-		allowedTools = append(allowedTools, tools.Tool{
-			Function: &tools.FunctionDefinition{
-				Name:        tool,
-				Description: tool,
-			},
-		})
+	for _, tool := range tls {
+		if slices.Contains(t.allowedTools, tool.Function.Name) {
+			allowedTools = append(allowedTools, tool)
+		}
 	}
+
 	return allowedTools, nil
 }
 

--- a/pkg/tools/builtin/filesystem_test.go
+++ b/pkg/tools/builtin/filesystem_test.go
@@ -852,3 +852,14 @@ func TestFilesystemTool_AddAllowedDirectory(t *testing.T) {
 		assert.Len(t, tool.allowedDirectories, 2)
 	})
 }
+
+func TestFilesystemTool_FilterTools(t *testing.T) {
+	allowedDirs := []string{"/tmp"}
+	tool := NewFilesystemTool(allowedDirs, WithAllowedTools([]string{"list_allowed_directories"}))
+
+	tools, err := tool.Tools(t.Context())
+	require.NoError(t, err)
+	require.Len(t, tools, 1)
+	require.Equal(t, "list_allowed_directories", tools[0].Function.Name)
+	require.NotNil(t, tools[0].Handler)
+}

